src/main/java/com/minecolonies/coremod/network/messages/server/SwitchBuildingWithToolMessage.java

```diff
@@ -0,0 +1,104 @@
+package com.minecolonies.coremod.network.messages.server;
+
+import com.ldtteam.structurize.items.ModItems;
+import com.minecolonies.api.network.IMessage;
+import net.minecraft.entity.player.ServerPlayerEntity;
+import net.minecraft.item.ItemStack;
+import net.minecraft.network.PacketBuffer;
+import net.minecraftforge.fml.LogicalSide;
+import net.minecraftforge.fml.network.NetworkEvent;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+/**
+ * Switch the buildtool with the respective item in the inventory.
+ */
+public class SwitchBuildingWithToolMessage implements IMessage
+{
+    /**
+     * The stack to switch.
+     */
+    private ItemStack stack;
+
+    /**
+     * Empty constructor used when registering the
+     */
+    public SwitchBuildingWithToolMessage()
+    {
+        super();
+    }
+
+    /**
+     * Switch the stack.
+     *
+     * @param stack the stack in the hand.
+     */
+    public SwitchBuildingWithToolMessage(final ItemStack stack)
+    {
+        super();
+        this.stack = stack;
+    }
+
+    /**
+     * Reads this packet from a {@link PacketBuffer}.
+     *
+     * @param buf The buffer begin read from.
+     */
+    @Override
+    public void fromBytes(@NotNull final PacketBuffer buf)
+    {
+        stack = buf.readItemStack();
+    }
+
+    /**
+     * Writes this packet to a {@link PacketBuffer}.
+     *
+     * @param buf The buffer being written to.
+     */
+    @Override
+    public void toBytes(@NotNull final PacketBuffer buf)
+    {
+        buf.writeItemStack(stack);
+    }
+
+    @Nullable
+    @Override
+    public LogicalSide getExecutionSide()
+    {
+        return LogicalSide.SERVER;
+    }
+
+    @Override
+    public void onExecute(final NetworkEvent.Context ctxIn, final boolean isLogicalServer)
+    {
+        final ServerPlayerEntity player = ctxIn.getSender();
+
+        int stackSLot = 0;
+        int buildToolSlot = 0;
+        for (int i = 0; i < 9; i++)
+        {
+            if (player.inventory.getStackInSlot(i).isItemEqual(stack))
+            {
+                stackSLot = i;
+            }
+            else if (player.inventory.getStackInSlot(i).getItem() == ModItems.buildTool.get())
+            {
+                buildToolSlot = i;
+            }
+        }
+
+        for (int i = 9; i < player.inventory.getSizeInventory(); i++)
+        {
+            if (player.inventory.getStackInSlot(i).getItem() == ModItems.buildTool.get())
+            {
+                buildToolSlot = i;
+            }
+        }
+
+        if (stackSLot != buildToolSlot)
```

Comment:
This looks like if it finds neither item then it will do nothing, but if it finds only the build tool it will swap that with whatever was in slot 0, and if it finds only the target stack then it will destroy whatever was in slot 0 and replace it with a build tool?  That doesn't sound right.
(Not that I'm entirely clear on the intended purpose of this message.)

Reply:
It shouldn't be possible for the stack to not exist, but true I'd better safeguard this.